### PR TITLE
🐛 (go/v4): return an errors when is not possible to update the resources instead of failing silently

### PR DESCRIPTION
**Pull request by onkar717**
## Description

`updateControllers` in `pkg/plugins/golang/options.go` discarded both calls to `Controllers.AddController` with `_ =`. `AddController` validates the name as a DNS-1035 label and rejects duplicates, but the error was never propagated anywhere.

`validateController` in `pkg/plugins/golang/v4/api.go` only checks emptiness/duplicates against the resource already persisted in the PROJECT file, and never runs its format check for a brand-new resource (`p.config.GetResource` returns an error for a resource that doesn't exist yet, so `validateController` returns `nil` immediately).

Net effect: for a new API, an invalid `--controller-name` (e.g. `Backup_Ctrl`, which fails DNS-1035 validation uppercase/underscore not allowed) was silently discarded. `res.Controllers` stayed empty, and the scaffolding logic in `api.go` fell back to the default kind-derived controller name instead, while the CLI reported success. The user's chosen
name was silently ignored.

Found this by reading the controller-scaffolding path end to end while looking for real, self-contained bugs to contribute. It's a genuine correctness gap (not defensive-only) reachable by any snake_case or otherwise-invalid `--controller-name` on first use, and the affected file  had zero test coverage for `ControllerName` at all beforehand.

## What this PR does

- `updateControllers` and `UpdateResource` (`pkg/plugins/golang/options.go`)  now return `error` and propagate both `AddController` calls instead of discarding them.
- Propagates that error through the three call sites (`pkg/plugins/golang/v4/api.go`, `pkg/plugins/golang/v4/webhook.go`,
  `pkg/plugins/golang/deploy-image/v1alpha1/api.go`), wrapped as `"error updating resource: %w"` to match the existing convention already used right below each of those call sites.
- Adds a `ControllerName validation` test block in `options_test.go`:  valid name (succeeds, controller added), invalid name (returns an error,
  `Controllers` stays empty instead of silently falling back), and duplicate name (returns an error).
- Updates the five pre-existing bare `options.UpdateResource(...)` calls in the same test file to assert `Succeed()`, since they now discard the same class of error the production code used to.

Verified locally: `go build ./...`, the `golang`/`v4`/`deploy-image`Ginkgo suites all pass, `golangci-lint run` (pinned v2.12.2, repo config) reports 0 issues. Also verified the new tests aren't vacuous by temporarily reverting the fix both the "invalid name" and "duplicate name" cases went red with the expected failure messages, then passed again after reverting
back.

Fixes: none filed this PR is the record; happy to file one if a maintainer would prefer it tracked separately.

### CODE REVIEW

**Review by camilamacedo86 — COMMENTED** [review]
/ok-to-test

File: pkg/plugins/golang/deploy-image/v1alpha1/api.go [hunk 1]
```diff
@@ -146,7 +146,9 @@ func (p *createAPISubcommand) InjectResource(res *resource.Resource) error {
 	p.options.DoController = true
 	p.options.Namespaced = true
 
-	p.options.UpdateResource(p.resource, p.config)
+	if err := p.options.UpdateResource(p.resource, p.config); err != nil {
+		return fmt.Errorf("error updating resource: %w", err)
+	}
 
 	if err := p.resource.Validate(); err != nil {
 		return fmt.Errorf("error validating resource: %w", err)
```
Comment (camilamacedo86) [inline pkg/plugins/golang/deploy-image/v1alpha1/api.go:154]: The description says `validateController` "only checked existing resources,
skipping validation for new ones", but the function is never changed here — the
`api.go` diff is only the wrap around `UpdateResource`. This early return is
what skips new resources, and it is still there:

```go
existingRes, err := p.config.GetResource(p.resource.GVK)
if err != nil {
    // Resource does not exist yet, no validation needed
    return nil
}

The PR routes around it by surfacing the error from AddController deep in the
data layer. Both changes are worth having, but the flag itself should be checked
where the other flags are checked:

// The name must be valid even when the resource is new and has nothing to compare against.
if p.options.ControllerName != "" {
    if err := (resource.Controller{Name: p.options.ControllerName}).Validate(); err != nil {
        return fmt.Errorf("invalid '--controller-name': %w", err)
    }
}

File: pkg/plugins/golang/options_test.go [hunk 2]
```diff
@@ -70,7 +70,7 @@ var _ = Describe("Options", func() {
 						Webhooks: &resource.Webhooks{},
 					}
 
-					options.UpdateResource(&res, cfg)
```
Comment (camilamacedo86) [inline pkg/plugins/golang/options_test.go:73]: The new tests exercise Options.UpdateResource directly. There's no test that the CLI path rejects create api --controller-name=Bad_Name, should we add that?

File: pkg/plugins/golang/deploy-image/v1alpha1/api.go [hunk 3]
```diff
@@ -146,7 +146,9 @@ func (p *createAPISubcommand) InjectResource(res *resource.Resource) error {
 	p.options.DoController = true
```
Comment (camilamacedo86) [inline pkg/plugins/golang/deploy-image/v1alpha1/api.go:146]: I think that 2/3 call  can never produce an error. webhook.go never sets DoController, so updateControllers sn't called at all. deploy-image/api.go sets DoController = true but never sets ControllerName, so it hits the early return nil. Only create api --controller-name=... can actually fail. Keeping the plumbing is fine, but the PR description should say so — otherwise a reviewer assumes webhook behavior changed.

Could you please check it out?

File: pkg/plugins/golang/v4/webhook.go [hunk 4]
```diff
@@ -165,7 +165,9 @@ func (p *createWebhookSubcommand) InjectResource(res *resource.Resource) error {
 		return errors.New("'--external-api-module' requires '--external-api-path' to be specified")
 	}
```
Comment (camilamacedo86) [inline pkg/plugins/golang/v4/webhook.go:159]: I think that to solve what you want to solve unless I misunderstood it we would need to:

Current:

```
      existingRes, err := p.config.GetResource(p.resource.GVK)
      if err != nil {
              // Resource does not exist yet, no validation needed
              return nil
      }
```

Change to:

```
      // The name must be rce is new and has nothingcompare against.
      if p.options.Control
              if err := (resource.Controller{Name: p.options.ControllerName}).
err != nil {
                      return fmt.Errorf("invalid '--controller-name': %w", err
              }
      }

      existingRes, err := p.config.GetResource(p.resource.GVK)
      if err != nil {
              // Resource does not exist yet, no further validation needed
              return nil
      }
```

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Hi @onkar717,

Thank you so much for looking into this.

I reviewed the changes, and I think we need to address a few points before merging. Could you please take another look?

I centralized my findings below and used AI to help organize the wording.

### 1. `validateController` still skips new resources

In `pkg/plugins/golang/v4/api.go`, the early return mentioned in the PR description is still unchanged:

```go
existingRes, err := p.config.GetResource(p.resource.GVK)
if err != nil {
    return nil // New resources are not validated.
}
```

Because `--controller-name` is a flag value, it should be validated before this return:

```go
if p.options.ControllerName != "" {
    if err := (resource.Controller{Name: p.options.ControllerName}).Validate(); err != nil {
        return fmt.Errorf("invalid '--controller-name': %w", err)
    }
}
```

### 2. `InjectResource` validates after mutating the configuration

In the same file, `UpdateResource` runs before `validateController`, so the deeper `AddController` error is always returned first.

The validation should happen before the configuration is mutated:

```go
if err := p.validateController(); err != nil {
    return err
}

if err := p.options.UpdateResource(p.resource, p.config); err != nil {
    return fmt.Errorf("error updating resource: %w", err)
}
```

`validateAPI()` can remain as the final call. `validateController` only reads the flags, GVK, and configuration, so this reordering should be safe.

Items 1 and 2 need to be addressed together. Applying only one of them would not change the current behavior.

### 3. Simplify the error message

The current error is:

```text
error updating resource: error adding controller "Backup_Ctrl": invalid controller name "Backup_Ctrl": ...
```

The controller name appears twice. `Controller.Validate()` already returns a complete error, so the wrapper only needs to identify the flag.

### 4. One discarded error remains

In `pkg/model/resource/resource.go:283`, the following error is still ignored:

```go
_ = r.Controllers.AddController(defaultName)
```

This follows the same legacy migration path. Since `Resource.Update` already returns an error, this should be a straightforward one-line fix.

This path is used when merging existing `PROJECT` file configurations.

### 5. Two call sites cannot currently trigger this error

`webhook.go` never sets `DoController`, and `deploy-image/api.go` never sets `ControllerName`.

At the moment, only `create api --controller-name` can trigger this validation error. Keeping the error propagation in the other call sites is fine, but the PR description should clarify this.

### 6. Add tests one layer higher

The new tests call `Options.UpdateResource` directly. Please also add:

* A CLI-level test confirming that `create api --controller-name=Bad_Name` is rejected.
* A test covering the legacy migration path in `resource.go`.

**Comment by camilamacedo86** [comment]
Hi @onkar717 

We use the PR title as release notes, I hope that you do not mind I tried to change it for users POV. Could you please check it out? if you think that has a better wording please feel free to change it too.

**Comment by camilamacedo86** [comment]
I think we need move with https://github.com/kubernetes-sigs/kubebuilder/pull/5941
Could you please give a look?
